core/src/main/scala/spinal/core/AFix.scala

```diff
@@ -760,183 +759,213 @@ class AFix(val maxRaw: BigInt, val minRaw: BigInt, val exp: Int) extends MultiDa
       res
     } else {
-      roundHalfDown(exp)
+      roundUp(exp, aligned)
     }
   }
 
-  def roundToInf(exp: Int, aligned: Boolean): AFix = roundHalfToInf(exp)
-
   /**
    * Rounds a value towards the nearest even value including half values, otherwise rounds towards odd values
    * @return Rounded result
    */
   def roundHalfToEven(exp: Int): AFix = {
+    roundToEven(exp, getTrunc.align)
+  }
+
+  def roundToEven(exp: Int, aligned: Boolean): AFix = {
     val drop = exp-this.exp
     if(drop < 0) return CombInit(this)
     val step = BigInt(1) << drop
-    val res = new AFix((this.maxRaw+step-1) >> drop, (this.minRaw+step-1) >> drop, exp)
+    val res = new AFix(((this.maxRaw+step-1) >> drop)-1, (this.minRaw+step-1) >> drop, exp)
 
     if (this.signed) {
-      res.raw := this.raw.asSInt.roundToEven(drop, false).resize(widthOf(res.raw)).asBits
+      res.raw := this.raw.asSInt.roundToEven(drop, aligned).resize(widthOf(res.raw)).asBits
     } else {
-      res.raw := this.raw.asUInt.roundToEven(drop, false).resize(widthOf(res.raw)).asBits
+      res.raw := this.raw.asUInt.roundToEven(drop, aligned).resize(widthOf(res.raw)).asBits
     }
 
     res
   }
 
-  def roundToEven(exp: Int, align: Boolean): AFix = roundHalfToEven(exp)
-
   /**
    * Rounds a value towards the nearest odd value including half values, otherwise rounds towards even values
    * @return Rounded result
    */
   def roundHalfToOdd(exp: Int): AFix = {
+    roundToOdd(exp, getTrunc.align)
+  }
+
+  def roundToOdd(exp: Int, align: Boolean): AFix = {
     val drop = exp-this.exp
     if(drop < 0) return CombInit(this)
     val step = BigInt(1) << drop
-    val res = new AFix((this.maxRaw+step-1) >> drop, (this.minRaw+step-1) >> drop, exp)
+    val res = new AFix(((this.maxRaw+step-1) >> drop)-1, (this.minRaw+step-1) >> drop, exp)
 
-    if (this.signed) {
-      res.raw := this.raw.asSInt.roundToOdd(drop, false).resize(widthOf(res.raw)).asBits
-    } else {
-      res.raw := this.raw.asUInt.roundToOdd(drop, false).resize(widthOf(res.raw)).asBits
+    if (this.signed) new AreaRoot{
+      res.raw := raw.asSInt.roundToOdd(drop, align).resize(widthOf(res.raw)).asBits
+    } else new AreaRoot {
+      res.raw := raw.asUInt.roundToOdd(drop, align).resize(widthOf(res.raw)).asBits
     }
 
     res
   }
 
-  def roundToOdd(exp: Int, align: Boolean): AFix = roundHalfToOdd(exp)
+  def getTrunc: TagAFixTruncated = {
+    this.getTag(classOf[TagAFixTruncated]).getOrElse(AFixTruncatedScope.get)
+  }
 
-  def round(exp: Int, aligned: Boolean): AFix = {
+  def round(exp: Int, aligned: Boolean = getTrunc.align): AFix = {
     val trunc = this.getTag(classOf[TagAFixTruncated])
 
     if (trunc.isDefined) {
-      this._round(trunc.get.rounding)
+      this._round(trunc.get.rounding, exp, aligned)
     } else {
-      roundHalfToInf(exp)
+      roundToInf(exp, aligned)
     }
   }
 
+  def fixTo(m: Int, exp: Int, roundType: RoundType): AFix = this.sat(m)._round(roundType, exp)
```

Comment:
This function seems dangerous, i mean, its signature is dangerous.
Int, Int, RoundType => could be used in other contextes

exp : ExpNumber  would already protect a bit, but still the first arguement being the number of bit that we would saturate seems arbitrary.

Why not instead doing myAFix.sat(x).fixTo(-8 exp, RoundingType.xxx) ?

Reply:
Yes, this signature is kind of misleading. Maybe we can rework it or just remove it?

Reply:
I would say let's remove that one, and update the few others which has exp to use the ExpNumber type